include/tvm/relax/expr.h

```diff
@@ -920,10 +920,16 @@ class FunctionNode : public BaseFuncNode {
   Expr body;
   /*! \brief The return type of the function. */
   StructInfo ret_struct_info;
+  /*! \brief Whether the function is annotated as pure or not. */
+  bool is_pure;
```

Comment:
Thanks for getting these in. One thing that I think would worth distinguish is purity annotation (which is more of a UX thing) and `force_pure`. I know it is getting into preference part :) So it is debatable.

- is_pure or purity annotation is something that I think is worth to be part of function annotation, since this is what we expect most users to annotate and is consistent with ret_struct_info
- force_pure on the other hand, is closer to attributes like `tir.noalias`, which affects the compiler internal, and might makes sense to remain as attribute `relax.force_pure` for now.

On the syntax side, would be nice if we can enable things like (can serve as a followup)
```
@R.function(pure=True, private=True)
def func(...)
```

Reply:
I tried to implement that but it gets really complicated because of how Python decorators work (namely, having optional arguments while still using `@R.function` without any explicit arguments). I'm sure it's probably possible but I wasn't able to get that working. Maybe it could be a separate decorator, like `@R.pure`.

Reply:
So would you say that I should have `force_pure` remain as an attribute, then?

Reply:
right, would be good to keep force_pure as an attr.

interesting, didn't think about that issue(optional argument while having R.function without explicit argument), perhaps allowing also a default for R.function can work. We can do that as a followup

Reply:
Implemented the change back to an attr